File: test/IECore/BlindDataHolder.py
##########################################################################
#
#  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

"""Unit test for BlindDataHolder binding"""
import os
import unittest

import IECore

class TestBlindDataHolder(unittest.TestCase):

	def testConstructors(self):
		"""Test BlindDataHolder constructors"""
		b = IECore.BlindDataHolder()

		c = IECore.CompoundData()
		c["floatData"] = IECore.FloatData(3.0)

		b = IECore.BlindDataHolder(c)

		self.assertEqual( b.typeName(), "BlindDataHolder" )
		self.failIf( IECore.Object.isAbstractType( "BlindDataHolder") )

	def testBlindData(self):
		"""Test BlindDataHolder blindData"""

		b = IECore.BlindDataHolder()

		b.blindData()["floatData"] = IECore.FloatData(1.0)
		b.blindData()["intData"] = IECore.IntData(-5)

		self.assertEqual( b.blindData()["floatData"].value, 1.0 )
		self.assertEqual( b.blindData()["intData"].value, -5 )

		self.assertEqual( len(b.blindData()), 2 )

	def testComparison(self):

		# test the empty case (where it doesn't allocate the compound data)
		a = IECore.BlindDataHolder( )
		b = IECore.BlindDataHolder( IECore.CompoundData() )
		c = IECore.BlindDataHolder( )
		c.blindData()

		self.assertEqual( a, a )
		self.assertEqual( a, a )
		self.assertEqual( a, b )
		self.assertEqual( b, a )
		self.assertEqual( a, c )
		self.assertEqual( c, a )
		self.assertEqual( b, c )
		self.assertEqual( c, b )

		c.blindData()['a'] = IECore.IntData(10)
		self.assertNotEqual( a, c )
		self.assertNotEqual( c, a )
		self.assertNotEqual( b, c )
		self.assertNotEqual( c, b )

	def testLoadSave(self):

		"""Test BlindDataHolder load/save"""

		iface = IECore.IndexedIO.create( "test/BlindDataHolder.fio", IECore.IndexedIO.OpenMode.Write )

		# first simple test: saving with some blind data
		b1 = IECore.BlindDataHolder()
		b1.blindData()["floatData"] = IECore.FloatData(1.0)
		b1.blindData()["intData"] = IECore.IntData(-5)
		b1.save( iface, "test" )

		b2 = IECore.Object.load( iface, "test" )
		self.assertEqual( b1, b2 )

		# should have written a "blindData" entry into the indexed io hierarchy
		self.failUnless( isinstance( iface.directory( ["test","data","BlindDataHolder", "data", "blindData"], IECore.IndexedIO.MissingBehaviour.NullIfMissing ), IECore.IndexedIO ) )

		# second test: overriding with no blind data
		b1 = IECore.BlindDataHolder()
		b1.save( iface, "test" )
		b1 = IECore.Object.load( iface, "test" )
		self.assertEqual( b1, IECore.BlindDataHolder() )

		# "blindData" entry should be excluded from the IndexedIO hierarchy
		self.assertEqual( iface.directory( ["test","data","BlindDataHolder"], IECore.IndexedIO.MissingBehaviour.NullIfMissing ), None )

	def testHash( self ) :

		b1 = IECore.BlindDataHolder()
		b2 = IECore.BlindDataHolder()

		self.assertEqual( b1.hash(), b2.hash() )

		b2.blindData()["a"] = IECore.FloatData( 1 )
		self.assertNotEqual( b1.hash(), b2.hash() )

	def tearDown(self):

		if os.path.isfile("./test/BlindDataHolder.fio") :
			os.remove("./test/BlindDataHolder.fio")

if __name__ == "__main__":
        unittest.main()

File: src/IECore/BlindDataHolder.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2013, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <cassert>
#include <iostream>

#include "IECore/BlindDataHolder.h"

using namespace IECore;

static IndexedIO::EntryID g_blindDataEntry("blindData");
const unsigned int BlindDataHolder::m_ioVersion = 1;

IE_CORE_DEFINEOBJECTTYPEDESCRIPTION(BlindDataHolder);

BlindDataHolder::BlindDataHolder()
{
	m_data = new CompoundData();
}

BlindDataHolder::BlindDataHolder(CompoundDataPtr data) : m_data(data)
{
	assert(m_data);
}

BlindDataHolder::~BlindDataHolder()
{
}

CompoundData *BlindDataHolder::blindData()
{
	return m_data.get();
}

const CompoundData *BlindDataHolder::blindData() const
{
	return m_data.get();
}

void BlindDataHolder::copyFrom( const Object *other, CopyContext *context )
{
	Object::copyFrom( other, context );
	const BlindDataHolder *tOther = static_cast<const BlindDataHolder *>( other );
	m_data = context->copy<CompoundData>( tOther->m_data.get() );
}

void BlindDataHolder::save( SaveContext *context ) const
{
	Object::save( context );

	bool haveData = m_data->readable().size();

	if ( haveData )
	{
		IndexedIOPtr container = context->container( staticTypeName(), m_ioVersion );
		context->save( m_data.get(), container.get(), g_blindDataEntry);
	}
}

void BlindDataHolder::load( LoadContextPtr context )
{
	Object::load( context );
	unsigned int v = m_ioVersion;
	IndexedIO::EntryID typeName = staticTypeName();

	ConstIndexedIOPtr container = context->container( typeName, v, false );
	if ( container )
	{
		m_data = context->load<CompoundData>( container.get(), g_blindDataEntry );
		assert(m_data);	
	}
	else
	{
		m_data = new CompoundData();
	}
}

bool BlindDataHolder::isEqualTo( const Object *other ) const
{
	if( !Object::isEqualTo( other ) )
	{
		return false;
	}
	const BlindDataHolder *tOther = static_cast<const BlindDataHolder *>( other );
	return m_data->isEqualTo( tOther->m_data.get() );
}

void BlindDataHolder::memoryUsage( Object::MemoryAccumulator &a ) const
{
	Object::memoryUsage( a );
	a.accumulate( m_data.get() );
}

void BlindDataHolder::hash( MurmurHash &h ) const
{
	Object::hash( h );
	// Our hash when blindData is empty or when m_data is unnitialized is the same.
	// This is currently garanteed by CompoundData but we are safer this way.
	if ( m_data->readable().size() )
	{
		m_data->hash( h );
	}
}
